### DISCUSSION TITLE
Grandparent Rollups & Ultimate Parent Rollups

### DISCUSSION THREAD

**Discussion by baobao917** [jamessimone/apex-rollup#504] [Q&A] [answered]
I've been reading the documentation on both grandparent (GP) vs ultimate parent (UP) rollups and trying to determine the correct one for my use case.  Below is my use case and questions:

Use Case:  I need to calculate the sum of Product Consumed.Total_Costs__c (child) up to the grandparent/ultimate parent of Assets.Total_Work_Order_Costs__c.  The path is from child (Product Consumed) > parent (Work Order) > grandparent/ultimate parent (Asset).  Therefore grandparent relationship field path is Work Order.Asset.Total_Work_Order_Costs__c.  

Also, if possible, we wish to bypass any sort of query/calculation on the intermediate object (Work Order) b/c we have lots of automation on that object and wish to minimize this calculation contributing to any SOQL limits.

1. Having a little difficulty understanding the difference between Grandparent rollups vs Ultimate Parent rollups. 

- In the UP rollup article, it states "it's not necessary for an ultimate parent rollup to be configured as a grandparent rollup unless you're also skipping intermediate objects. If you are skipping intermediate objects, make sure to follow the steps outlined in the Grandparent Rollup wiki page, prior to proceeding here."

- In the GP rollup article, it states "all intermediate objects (Opportunity, in our example) must also have a RT-flow setup to call the Apex Rollup actions. Again, this is to ensure rollups are always calculated correctly if any of the intermediate parent objects have their lookup fields changed."  

I've interpreted this to mean that GP rollups skips intermediate objects (1st bullet above) but in 2nd bullet I've interpreted it to mean it doesn't skip intermediate objects.  I thought that both GP and UP rollups perform rollup calculations from the child object directly to the GP/UP object without involving the intermediate objects.  Could you please explain more simply what's the difference between the two?  

2. As stated in our use case above, we wish to bypass any sort of query/calculation on the intermediate object (Work Order) due to SOQL query limits we are running up against whenever the work order object is involved.  In the GP article, it states that all intermediate objects must also have a flow/apex trigger to ensure rollups are calculated correctly if any of the intermediate parent objects have their lookup fields changed.  Is that necessary when the child & intermediate object exists as a master-detail relationship (Work Order-Product Consumed)?
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/076bc5bc-b682-4bbe-9892-35a85d139e2d)

3. Given the use case stated above, which is the recommended approach? GP vs UP rollups?

**Comment by jamessimone** [comment, marked as answer]
Apologies for the confusion. You don't need an ultimate parent rollup for this - that's for when you're rolling up on a self-referential (hierarchy) field like Parent Account. The basic grandparent rollup you described should suffice.

You only need to call Apex Rollup on the intermediate objects for situations where reparenting the intermediate objects wouldn't recall automation on the children; in your case, I think you should be fine as is since it's a true master detail. I will update the README and associated docs to make this clearer

**Comment by baobao917** [comment]
Thanks @jamessimone !  One quick follow-up....so in a GP rollup, will it bypass the intermediate object altogether? (i.e. won't perform unnecessary SOQL queries against the work order object)

**Reply by jamessimone** [reply]
@baobao917 it will query work order but that shouldn't matter because those records would never be locked and won't have DML performed on them
